### ISSUE TITLE
VideoView Pixelation on External Monitor

### ISSUE THREAD

**Issue by i-costello** [twilio/twilio-video-ios#123]
### Description

I am trying to display participant video on both the iPhone and on an external monitor, connected via Screen Mirroring, simultaneously. The external monitor is managed by an additional UIScene, which has its own view hierarchy.

The external scene has a VideoView for the local participant’s video and another for a remote participant’s video. In both views, the video content is pixelated and has jagged edges. Additionally, the video content is squished vertically, and does not fill the VideoView. Setting the content mode to scaleAspectFill or scaleAspectFit fills the VideoView, but does not correct the distorted aspect ratio.

### Steps to Reproduce

1. Set up an external scene by adding support in the target’s Info.plist file and creating an additional UIWindowSceneDelegate subclass (e.g. ExtSceneDelegate) with its own root view controller. 
```swift
<key>UIApplicationSceneManifest</key>
<dict>
	<key>UIApplicationSupportsMultipleScenes</key>
	<true/>
	<key>UISceneConfigurations</key>
	<dict>
		<key>UIWindowSceneSessionRoleApplication</key>
		<array>
			<dict>
				<key>UISceneConfigurationName</key>
				<string>Default Configuration</string>
				<key>UISceneDelegateClassName</key>
				<string>$(PRODUCT_MODULE_NAME).SceneDelegate</string>
			</dict>
		</array>
		<key>UIWindowSceneSessionRoleExternalDisplay</key>
		<array>
			<dict>
				<key>UISceneConfigurationName</key>
				<string>External Screen</string>
				<key>UISceneDelegateClassName</key>
				<string>$(PRODUCT_MODULE_NAME).ExtSceneDelegate</string>
			</dict>
		</array>
	</dict>
</dict>
```

2. Add a VideoView to the external view hierarchy, and add that view as a renderer for a video track (make sure this is done after the external monitor is connected).
3. To view the external scene: From the Control Center, tap Screen Mirroring and select an Apple TV or other device that supports mirroring (tested on Apple TV and a MacBook Pro using Reflector 3). 

#### Expected Behavior

The video content is rendered at the same quality and aspect ratio on the external monitor as it is rendered on the iPhone.

#### Actual Behavior

The video content is rendered at a distorted aspect ratio (squished vertically) on the external monitor. The aspect ratio can be corrected by setting width and height layout constraints matching the video content’s aspect ratio for the VideoView, and setting the contentMode to scaleToFill. However, the content is then pixelated/ shown at a lower quality than the other renderer on the iPhone.

![VideoView Pixelation](https://user-images.githubusercontent.com/28639532/90614533-a1561d80-e1bf-11ea-9e64-c8225dc4c231.png)

#### Reproduces How Often

Every time

### Versions

All relevant version information for the issue.

#### Video iOS SDK

3.5.0 via CocoaPods

#### Xcode

11.6

#### iOS Version

13.x.x

#### iOS Device

iPhone XR

**Comment by ceaglest** [comment]
Hi @i-costello,

Thanks for the report, we will need to reproduce this locally before commenting much further.

Typically, the MTKView sets the contentScaleFactor automatically when it transitions to a new window but it doesn't appear that this is working properly.

Best,
Chris

**Comment by piyushtank** [comment]
@i-costello What you found might be an underlying configuration bug in our Metal Renderer Pipeline. We have never tried Screen Mirroring in our lab locally. And at present we are all busy working on upgrading our media engine (webrtc) upgrade in 4.0 release. We will try this out in future. I will keep you posted.

If you are blocked, you can try our deprecated APIs of using OpenGL renderer pipeline to see if that solves the problem. Please note that we plan to remove Open GL renderer APIs in our 4.0 release, so I wouldn't use deprecated APIs unless it is absolutely necessary.

**Comment by piyushtank** [comment]
@i-costello I tried using Reflactor 3 (trial mode), and following is what I observed. 

On phone screen -

![OnPhoneScreen](https://user-images.githubusercontent.com/17037593/101112343-ec8b1500-3591-11eb-887d-fca7354e1ac9.PNG)

Mirrored screen- 

<img width="858" alt="OnMacBookWithReflactor3" src="https://user-images.githubusercontent.com/17037593/101112354-f14fc900-3591-11eb-84ce-3d31a8f6cd43.png">

I could not reproduce the problem. When the video resolution is downscaled during the call, it is possible when the screen is enlarged to bigger screen, the artifacts are more apparent. Can you confirm if you are observing the issue for LocalVideo track as well?

**Comment by piyushtank** [comment]
@i-costello Finally I was able to reproduce the problem. Thanks for reporting the bug. We have just shipped 4.0.1 SDK with the fix for this issue. [Here](https://www.twilio.com/docs/video/changelog-twilio-video-ios-latest#401-december-18-2020) are the changelogs. Try it out and let us know if you have any questions.

Best,
Piyush
